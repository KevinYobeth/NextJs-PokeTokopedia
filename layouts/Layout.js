import Navbar from '../components/Navbar'

export default function Layout({ children }) {
  return (
    <>
      <Navbar />
      <main
        className='pt-24 mx-auto 
        sm:max-w-screen-sm 
        md:max-w-screen-md 
        lg:max-w-screen-lg 
        px-5'
      >{children}</main>
    </>
  );
}