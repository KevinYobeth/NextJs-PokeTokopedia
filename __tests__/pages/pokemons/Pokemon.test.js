import { render, screen, act, fireEvent } from '@testing-library/react';
import { mockCatchedPokemons, mockPokemonDetails, mockPokemonDetailsResponse }
  from '../../../fixtures/PokemonFixtures';
import Layout from '../../../layouts/Layout';
import Pokemon, { getServerSideProps } from '../../../pages/pokemons/[name]';
import App, { PokemonContext } from '../../../pages/_app';
import axios from 'axios';

jest.mock('axios');
jest.spyOn(window.localStorage.__proto__, 'setItem');
window.localStorage.__proto__.setItem = jest.fn();

describe('Pokemon', () => {
  const renderComponent = (catchedPokemons = [], handleAddPokemon = null,
    thumbnail, name, id, pokemonDetails) => render(
      <PokemonContext.Provider value={{ catchedPokemons, handleAddPokemon }}>
        <Layout>
          <Pokemon thumbnail={thumbnail} name={name} id={id} pokemonDetails={pokemonDetails} />
        </Layout>
      </PokemonContext.Provider>
    );

  const renderComponentWithActualContext = (thumbnail, name, id, pokemonDetails) => {
    const props = { thumbnail, name, id, pokemonDetails };
    return render(
      <App Component={Pokemon} pageProps={{ ...props }} />
    );
  };

  describe('#render', () => {
    it('should render Pokemon pages correctly when loaded', () => {
      const { thumbnail, name, id, pokemonDetails } = mockPokemonDetails;
      const { container } = renderComponent(mockCatchedPokemons, null,
        thumbnail, name, id, pokemonDetails);

      expect(container).toMatchSnapshot();
    });
  });

  describe('#onClick', () => {
    it('should fail to catch pokemon and show fail message when button is invoked', async () => {
      const mockMath = Object.create(global.Math);
      mockMath.random = () => 0.3;
      global.Math = mockMath;
      const { thumbnail, name, id, pokemonDetails } = mockPokemonDetails;
      renderComponent(mockCatchedPokemons, null,
        thumbnail, name, id, pokemonDetails);
      const catchPokemonButton = screen.getByText('Catch charmander');

      await act(async () => {
        await catchPokemonButton.click();
      });

      expect(screen.getByText('Sorry you failed to catch charmander')).toBeInTheDocument();
    });

    it('should succeed to catch pokemon and show success message when button is invoked',
      async () => {
        const mockMath = Object.create(global.Math);
        mockMath.random = () => 0.8;
        global.Math = mockMath;
        const { thumbnail, name, id, pokemonDetails } = mockPokemonDetails;
        renderComponent(mockCatchedPokemons, null,
          thumbnail, name, id, pokemonDetails);
        const catchPokemonButton = screen.getByText('Catch charmander');

        await act(async () => {
          await catchPokemonButton.click();
        });

        expect(screen.getByText('Congratulations! You catch charmander')).toBeInTheDocument();
      });

    it('should succeed to catch pokemon and user should be able to add pokemon to collection when'
      + 'invoked', async () => {
        const mockMath = Object.create(global.Math);
        mockMath.random = () => 0.8;
        global.Math = mockMath;
        const mockHandleAddPokemon = jest.fn();
        const { thumbnail, name, id, pokemonDetails } = mockPokemonDetails;
        renderComponent(mockCatchedPokemons, mockHandleAddPokemon,
          thumbnail, name, id, pokemonDetails);
        const catchButton = screen.getByText('Catch charmander');

        await act(async () => {
          await catchButton.click();
        });
        await act(async () => {
          const addButton = screen.getByText('Add to Collection');
          const nicknameInput = screen.getByTestId('nickname-input');
          await fireEvent.change(nicknameInput, { target: { value: 'budi' } });
          await addButton.click();
        });
        const ownedCount = screen.getByTestId('owned-count');

        expect(screen.getByText('budi is added to your collection!')).toBeInTheDocument();
        expect(ownedCount).toHaveTextContent('2');
        expect(mockHandleAddPokemon)
          .toHaveBeenCalledWith({ name, id, thumbnail, nickname: 'budi' });
      });

    it('should succeed to catch pokemon and user should be able to add pokemon to collection using'
      + 'enter key', async () => {
        const mockMath = Object.create(global.Math);
        mockMath.random = () => 0.8;
        global.Math = mockMath;
        const mockHandleAddPokemon = jest.fn();
        const { thumbnail, name, id, pokemonDetails } = mockPokemonDetails;
        renderComponent(mockCatchedPokemons, mockHandleAddPokemon,
          thumbnail, name, id, pokemonDetails);
        const catchButton = screen.getByText('Catch charmander');

        await act(async () => {
          await catchButton.click();
        });
        await act(async () => {
          const nicknameInput = screen.getByTestId('nickname-input');
          await fireEvent.change(nicknameInput, { target: { value: 'bambang' } });
          await fireEvent.keyPress(nicknameInput, { key: 'Enter', code: 'Enter', charCode: 13 });
        });

        expect(mockHandleAddPokemon)
          .toHaveBeenCalledWith({ name, id, thumbnail, nickname: 'bambang' });
        expect(screen.getByText('bambang is added to your collection!')).toBeInTheDocument();
      });

    it('should succeed to catch pokemon and fail validation when given duplicate or empty name',
      async () => {
        const mockMath = Object.create(global.Math);
        mockMath.random = () => 0.8;
        global.Math = mockMath;
        const { thumbnail, name, id, pokemonDetails } = mockPokemonDetails;
        renderComponent(mockCatchedPokemons, null,
          thumbnail, name, id, pokemonDetails);
        const catchButton = screen.getByText('Catch charmander');

        await act(async () => {
          await catchButton.click();
        });
        await act(async () => {
          const addButton = screen.getByText('Add to Collection');
          const nicknameInput = screen.getByTestId('nickname-input');
          await fireEvent.change(nicknameInput, { target: { value: '' } });
          await addButton.click();
          await fireEvent.change(nicknameInput, { target: { value: 'kevin' } });
          await addButton.click();
        });

        expect(screen.getByText('Nickname must be unique!')).toBeInTheDocument();
      });

    it('should show relevant tab informations when tab category title is clicked', async () => {
      const { thumbnail, name, id, pokemonDetails } = mockPokemonDetails;
      renderComponent(mockCatchedPokemons, null,
        thumbnail, name, id, pokemonDetails);
      const abilitiesTab = screen.getByText('Abilities');
      const movesTab = screen.getByText('Moves');
      const detailsTab = screen.getByText('Details');

      await act(async () => {
        await movesTab.click();
        await detailsTab.click();
        await abilitiesTab.click();
      });

      expect(screen.getByText('Blaze')).toBeInTheDocument();
    });

    it('should call handleAddPokemon from app.js when pokemon is catched', async () => {
      const mockMath = Object.create(global.Math);
      mockMath.random = () => 0.8;
      global.Math = mockMath;
      const { thumbnail, name, id, pokemonDetails } = mockPokemonDetails;
      renderComponentWithActualContext(thumbnail, name, id, pokemonDetails);
      const catchButton = screen.getByText('Catch charmander');

      await act(async () => {
        await catchButton.click();
      });
      await act(async () => {
        const nicknameInput = screen.getByTestId('nickname-input');
        await fireEvent.change(nicknameInput, { target: { value: 'bambang' } });
        await fireEvent.keyPress(nicknameInput, { key: 'Enter', code: 'Enter', charCode: 13 });
      })

      expect(localStorage.setItem).toHaveBeenCalledWith('catchedPokemons',
        JSON.stringify([{ name, id, nickname: 'bambang', thumbnail }]));
    });
  });

  describe('#getServerSideProps', () => {
    it('should call pokemon API to fetch data when page loaded', async () => {
      axios.get.mockResolvedValue({ data: mockPokemonDetailsResponse });

      const response = await getServerSideProps({ params: { name: 'charmander' } });

      expect(axios.get).toHaveBeenCalledWith('https://pokeapi.co/api/v2/pokemon/charmander');
      expect(response).toEqual({ props: mockPokemonDetails });
    });
  });
});