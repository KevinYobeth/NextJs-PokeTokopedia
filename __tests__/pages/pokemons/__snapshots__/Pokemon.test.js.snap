// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon #render should render Pokemon pages correctly when loaded 1`] = `
<div>
  <div
    class="fixed flex items-center justify-center w-full h-20 bg-white drop-shadow-lg 
 z-30"
  >
    <div
      class="flex items-center justify-between mx-auto w-full p-5 sm:max-w-screen-sm
 md:max-w-screen-md lg:max-w-screen-lg"
    >
      <a
        href="/"
      >
        <h1
          class="text-4xl font-bold"
        >
          POKE
        </h1>
      </a>
      <a
        data-testid="pokeball"
        href="/pokemons"
      >
        <div
          class="flex items-center gap-x-2"
        >
          <p>
            (
            3
            )
          </p>
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDAiIGhlaWdodD0iNDAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="my pokemon"
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
      </a>
    </div>
  </div>
  <main
    class="pt-24 mx-auto 
 sm:max-w-screen-sm 
 md:max-w-screen-md 
 lg:max-w-screen-lg 
 px-5"
  >
    <div
      class="flex flex-col md:flex-row gap-5"
    >
      <div
        class="flex flex-col self-start justify-self-start w-full md:w-2/5 bg-white 
 rounded-md drop-shadow-md p-5 gap-y-3"
      >
        <div
          class="-mt-4"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTEyIiBoZWlnaHQ9IjUxMiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="charmander"
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
        <div
          class="flex flex-col w-full gap-y-5 text-center"
        >
          <div
            class="flex flex-col text-center"
          >
            <h2
              class="font-poppins font-medium text-3xl"
            >
              Charmander
              #
              4
            </h2>
            <p
              class="font-poppins text-sm"
              data-testid="owned-count"
            >
              Owned: 
              <span>
                2
              </span>
            </p>
          </div>
          
          <div
            class="flex flex-col gap-y-2"
          >
            <button
              class="w-full rounded-md p-3 bg-gray-800 hover:bg-gray-900 text-white -translate-y-1.5 
      hover:-translate-y-2.5 active:-translate-y-1 duration-300 active:duration-75 
      transition-transform"
            >
              <p
                class="font-poppins text-center"
              >
                Catch charmander
              </p>
            </button>
            <a
              class="w-full"
              href="/"
            >
              <button
                class="w-full rounded-md p-3 bg-gray-300 hover:bg-gray-200 text-black -translate-y-1.5 
      hover:-translate-y-2.5 active:-translate-y-1 duration-300 active:duration-75 
      transition-transform"
              >
                <p
                  class="font-poppins text-center"
                >
                  View All Pokemons
                </p>
              </button>
            </a>
          </div>
        </div>
      </div>
      <div
        class="flex flex-col"
      >
        <div
          class="flex font-poppins font-medium text-lg w-full gap-x-10 md:mt-0 cursor-pointer"
        >
          <button
            class="text-black"
          >
            Details
          </button>
          <button
            class="text-gray-400"
          >
            Moves
          </button>
          <button
            class="text-gray-400"
          >
            Abilities
          </button>
        </div>
        <div
          class="flex flex-col mt-5"
        >
          <div
            class="flex"
          >
            <div
              class="grid grid-cols-2 w-full"
            >
              <p
                class="font-poppins font-medium text-gray-400 col-span-1"
              >
                Id
              </p>
              <p
                class="font-poppins font-medium text-black col-span-1"
              >
                4
              </p>
            </div>
          </div>
          <div
            class="flex"
          >
            <div
              class="grid grid-cols-2 w-full"
            >
              <p
                class="font-poppins font-medium text-gray-400 col-span-1"
              >
                Name
              </p>
              <p
                class="font-poppins font-medium text-black col-span-1"
              >
                Charmander
              </p>
            </div>
          </div>
          <div
            class="flex"
          >
            <div
              class="grid grid-cols-2 w-full"
            >
              <p
                class="font-poppins font-medium text-gray-400 col-span-1"
              >
                Height
              </p>
              <p
                class="font-poppins font-medium text-black col-span-1"
              >
                6 decimetres
              </p>
            </div>
          </div>
          <div
            class="flex"
          >
            <div
              class="grid grid-cols-2 w-full"
            >
              <p
                class="font-poppins font-medium text-gray-400 col-span-1"
              >
                Weight
              </p>
              <p
                class="font-poppins font-medium text-black col-span-1"
              >
                85 hectograms
              </p>
            </div>
          </div>
          <div
            class="flex"
          >
            <div
              class="grid grid-cols-2 w-full"
            >
              <p
                class="font-poppins font-medium text-gray-400 col-span-1"
              >
                Types
              </p>
              <p
                class="font-poppins font-medium text-black col-start-2 col-span-1"
              >
                Fire
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
